### ISSUE TITLE
 [master] 'openssl dgst' can't access digests that libcrypto doesn't know about

### ISSUE THREAD

**Issue by levitte** [openssl/openssl#14179]
I played a bit with external providers that offer digests that libcrypto doesn't know about, and that fails.
This is due to opt_md() (in apps/lib/opt.c) only tries to get the cipher with EVP_get_digestbyname(), never with EVP_MD_fetch().

**Comment by levitte** [comment]
In #14182, a few solutions were discussed, I'll repeat them here

**Comment by levitte** [comment]
Solution 1, an apps hack (from https://github.com/openssl/openssl/pull/14182#discussion_r575867943):

``` diff
diff --git a/apps/lib/opt.c b/apps/lib/opt.c
index 8cc520daec..becf46f11b 100644
--- a/apps/lib/opt.c
+++ b/apps/lib/opt.c
@@ -353,9 +353,12 @@ void print_format_error(int format, unsigned long flags)
 }
 
 /* Parse a cipher name, put it in *EVP_CIPHER; return 0 on failure, else 1. */
-int opt_cipher(const char *name, const EVP_CIPHER **cipherp)
+int opt_cipher(const char *name,
+               const EVP_CIPHER **cipherp, EVP_CIPHER **fetched_cipherp)
 {
-    *cipherp = EVP_get_cipherbyname(name);
+    if ((*cipherp = EVP_get_cipherbyname(name)) != NULL
+        || (*cipherp = *fetched_cipherp =
+            EVP_CIPHER_fetch(NULL, name, NULL)) != NULL)
     if (*cipherp != NULL)
         return 1;
     opt_printf_stderr("%s: Unknown cipher: %s\n", prog, name);
@@ -365,9 +368,11 @@ int opt_cipher(const char *name, const EVP_CIPHER **cipherp)
 /*
  * Parse message digest name, put it in *EVP_MD; return 0 on failure, else 1.
  */
-int opt_md(const char *name, const EVP_MD **mdp)
+int opt_md(const char *name, const EVP_MD **mdp, EVP_MD **fetched_mdp)
 {
-    *mdp = EVP_get_digestbyname(name);
+    if ((*mdp = EVP_get_digestbyname(name)) != NULL
+        || (*mdp = *fetched_mdp =
+            EVP_MD_fetch(NULL, name, NULL)) != NULL)
     if (*mdp != NULL)
         return 1;
     opt_printf_stderr("%s: Unknown option or message digest: %s\n", prog,
```

Of course, it means that every caller needs to adapt.  That's not so bad, here's the result in `apps/enc.c`:

``` diff
diff --git a/apps/enc.c b/apps/enc.c
index 9982337c01..dfb5c239ad 100644
--- a/apps/enc.c
+++ b/apps/enc.c
@@ -109,8 +110,8 @@ int enc_main(int argc, char **argv)
     BIO *in = NULL, *out = NULL, *b64 = NULL, *benc = NULL, *rbio =
         NULL, *wbio = NULL;
     EVP_CIPHER_CTX *ctx = NULL;
-    const EVP_CIPHER *cipher = NULL;
-    const EVP_MD *dgst = NULL;
+    const EVP_CIPHER *cipher = NULL; EVP_CIPHER *fetched_cipher = NULL;
+    const EVP_MD *dgst = NULL; EVP_MD *fetched_dgst = NULL;
     const char *digestname = NULL;
     char *hkey = NULL, *hiv = NULL, *hsalt = NULL, *p;
     char *infile = NULL, *outfile = NULL, *prog;
@@ -297,7 +303,7 @@ int enc_main(int argc, char **argv)
 
     /* Get the cipher name, either from progname (if set) or flag. */
     if (ciphername != NULL) {
-        if (!opt_cipher(ciphername, &cipher))
+        if (!opt_cipher(ciphername, &cipher, &fetched_cipher))
             goto opthelp;
     }
     if (cipher && EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_AEAD_CIPHER) {
@@ -309,7 +315,7 @@ int enc_main(int argc, char **argv)
         goto end;
     }
     if (digestname != NULL) {
-        if (!opt_md(digestname, &dgst))
+        if (!opt_md(digestname, &dgst, &fetched_dgst))
             goto opthelp;
     }
     if (dgst == NULL)
@@ -625,6 +634,8 @@ int enc_main(int argc, char **argv)
         BIO_printf(bio_err, "bytes written: %8ju\n", BIO_number_written(out));
     }
  end:
+    EVP_CIPHER_free(fetched_cipher);
+    EVP_MD_free(fetched_dgst);
     ERR_print_errors(bio_err);
     OPENSSL_free(strbuf);
     OPENSSL_free(buff);
```

**Comment by levitte** [comment]
Solution 2 (from https://github.com/openssl/openssl/pull/14182#issuecomment-778993272):

``` diff
diff --git a/crypto/evp/digest.c b/crypto/evp/digest.c
index 3dfcfcda8e..1696d3fb11 100644
--- a/crypto/evp/digest.c
+++ b/crypto/evp/digest.c
@@ -1017,7 +1017,7 @@ void EVP_MD_free(EVP_MD *md)
 {
     int i;
 
-    if (md == NULL)
+    if (md == NULL || md->prov == NULL)
         return;
 
     CRYPTO_DOWN_REF(&md->refcnt, &i, md->lock);
diff --git a/crypto/evp/names.c b/crypto/evp/names.c
index cb59813857..1c8c84f741 100644
--- a/crypto/evp/names.c
+++ b/crypto/evp/names.c
@@ -105,29 +105,29 @@ const EVP_CIPHER *evp_get_cipherbyname_ex(OSSL_LIB_CTX *libctx,
 
 static void digest_from_name(const char *name, void *data)
 {
-    const EVP_MD **md = data;
+    EVP_MD **md = data;
 
     if (*md != NULL)
         return;
 
-    *md = (const EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);
+    *md = (EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);
 }
 
-const EVP_MD *EVP_get_digestbyname(const char *name)
+EVP_MD *EVP_get_digestbyname(const char *name)
 {
     return evp_get_digestbyname_ex(NULL, name);
 }
 
-const EVP_MD *evp_get_digestbyname_ex(OSSL_LIB_CTX *libctx, const char *name)
+EVP_MD *evp_get_digestbyname_ex(OSSL_LIB_CTX *libctx, const char *name)
 {
-    const EVP_MD *dp;
+    EVP_MD *dp;
     OSSL_NAMEMAP *namemap;
     int id;
 
     if (!OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_DIGESTS, NULL))
         return NULL;
 
-    dp = (const EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);
+    dp = (EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);
 
     if (dp != NULL)
         return dp;
diff --git a/include/openssl/evp.h b/include/openssl/evp.h
index 5f9de9d8b9..7658a90c1a 100644
--- a/include/openssl/evp.h
+++ b/include/openssl/evp.h
@@ -1100,7 +1100,7 @@ int EVP_add_cipher(const EVP_CIPHER *cipher);
 int EVP_add_digest(const EVP_MD *digest);
 
 const EVP_CIPHER *EVP_get_cipherbyname(const char *name);
-const EVP_MD *EVP_get_digestbyname(const char *name);
+EVP_MD *EVP_get_digestbyname(const char *name);
 
 void EVP_CIPHER_do_all(void (*fn) (const EVP_CIPHER *ciph,
                                    const char *from, const char *to, void *x),
```

**Comment by levitte** [comment]
I have tested both solutions, and I can squarely say that:

-   Solution 1 works!  I have tested it, among others with the budding GOST provider (gost-engine/engine#313)
-   Solution 2...  has problems, which aren't easy to resolve.  I'll talk more about it in the next comment.

**Comment by mattcaswell** [comment]
>  have tested it, among others with the budding GOST provider

A GOST Provider!! Nice!

**Comment by beldmit** [comment]
> > have tested it, among others with the budding GOST provider
> 
> A GOST Provider!! Nice!

Very impressive work! 
But for the testing purposes it makes sense to have both engine and provider.

**Comment by levitte** [comment]
So, problems with Solution 2:

We need to keep track of how the EVP_MD was created.  There are three possibilities:

1.  It's a constant from within libcrypto.  That should never be freed
2.  It was created with EVP_MD_meth_new().  This classically happens in ENGINEs, but may also be done with applications.  Everyone else should regard it a constant, just like (1), and it only needs to be freed once when not needed any more, using EVP_MD_meth_free().
3.  It was created with EVP_MD_fetch().  This is a provider backed method, and is reference counted, i.e. is *not* regarded as constant by others, and if explicitly fetched, it must also be explicitly freed using EVP_MD_free()

What I've worked out is that a flag can help determine if EVP_MD_meth_new() created the EVP_MD, and thusly if EVP_MD_meth_free() should be allowed to free it.
Likewise, we can use the OSSL_PROVIDER pointer in the EVP_MD to determine if EVP_MD_free() is allowed to free it (EVP_MD_free() is a *new* function, so it can be relatively safe to say that it should only free provider backed EVP_MDs).

You all with me so far?  Ok, so let's look at this line:

https://github.com/openssl/openssl/blob/9dc9c7f2d7f25e100dd7d80e9bc51e205033cd8c/apps/dgst.c#L406

One would think that this is easy, just cast the return value to `(EVP_MD *)` and be happy.  But then what?  Thing is, if *md* points at an explicitly fetched `EVP_MD`, it should free it explicitly with `EVP_MD_free()`, or we will have a leak because the refcount wasn't decremented enough.
The solution to that problem could, of course, be for `EVP_MD_CTX_md()` to increment the refcount if the `EVP_MD` returned is provider backed, i.e. view that as an explicit fetch.  However, here's the catch: older applications assume that `EVP_MD_CTX_md()` return a constant, and have therefore no `EVP_MD_free()` call.  So if they get a provider backed `EVP_MD`, that will cause them to leak.
We could also continue to have `EVP_MD_CTX_md()` just return the `EVP_MD` and assume that it's going to be treated as a constant.  However, if the application does a naïve `md = (EVP_MD *)EVP_MD_CTX_md()` and later on calls `EVP_MD_free(md)`, that will lead to a double free.

This is a bad case of catch-22 with Solution 2.

From this experience, I would say that it's still better to go for solution 1, and require that applications keep track of what they fetch explicitly and what they don't, to avoid getting tangled in an impossible or *very* delicate situation.

**Comment by levitte** [comment]
> But for the testing purposes it makes sense to have both engine and provider.

Yes, and as you can see, that PR does exactly that.
But, let's not derail this issue, yeah?

**Comment by richsalz** [comment]
I disagree with your catch-22 reasoning (if only because a flag can be three-way), but since you already did something in #14194 it's not worth spending time discussing.

**Comment by levitte** [comment]
Well, I actually tried...  and got into double free errors, or a memleak, all depending on having or not having an EVP_MD_free() at the end.  Whatever you do that's *uniform* (i.e. not having to have to variables for distinction), it risks crashing or having your sanitizer scream (and oh boy is it loud...).

So yeah, if someone can *actually* work, without having fragile corner cases, I'm ready to listen.  In the mean time, the *experience* I gained within an hour wasn't very encouraging.

**Comment by t8m** [comment]
@levitte - the only sane way how the md returned from EVP_MD_CTX_md(tctx) could be handled is to have the returned md as const and treat it as if the call is EVP_MD_CTX_get0_md(). So yeah you really cannot mix md returned from EVP_MD_CTX_md with md returned from explicit fetch.

Yeah, previously all md objects were basically treated like const (ignoring the EVP_MD_meth_new() case) so the md returned from EVP_getdigestbyname() and EVP_MD_CTX_md() could be treated the same. If we introduced the flag you could simplify your app by calling EVP_MD_free() on md returned from the EVP_getdigestbyname() but never on a md returned from EVP_MD_CTX_md(), that would have to be handled separately.

**Comment by richsalz** [comment]
Is the "catch-22" issue only for MessageDigest, or does it also happen for ciphers?

**Comment by levitte** [comment]
> Is the "catch-22" issue only for MessageDigest, or does it also happen for ciphers?

Essentially, it comes down to a base principle related to explicit fetches.  Any application (or library for that matter) that does an *explicit fetch* must also do an *explicit free* of that pointer.  `foo = EVP_MD_fetch()` is an explicit fetch, `foo = EVP_get_digestbyname()` or `foo = EVP_MD_CTX_md()` are not.

So what this essentially boils down to is that applications (or libraries for that matter) *must* keep track of their fetched pointers, and free those at some point, but *only* those.  Applications that mix in engines need to handle the fact that methods coming from engines (for example, through a call of `EVP_get_digestbyname()`) must *not* be freed.  The easiest way to handle the combination is to do this for engine methods (i.e. the old way):

``` C
    const EVP_MD *md = EVP_get_digestbyname("whatever");
```

and do this when fetching:

``` C
    EVP_MD *fetched_md;
    const EVP_MD *md = fetched_md = EVP_MD_fetch("whatever");
```

In both cases, `md` can be used to perform the actual operations, and libcrypto will do its *implicit* fetch if it gets the pointer that came from `EVP_get_digestbyname()`.  All the application needs to add somewhere is this:

``` C
    EVP_MD_free(fetched_md);
```

In other words, `fetched_md` *only* needs to be used to keep track of explicit fetches that need to be freed.

The distinction between implicit and explicit fetches is documented, including the fact the the code doing the explicit fetch must also free it.

**Comment by richsalz** [comment]
Again: is it only `EVP_MD_free` that has this problem?  I suppose I can test it by doing something like:
```
EVP_CIPH_free(EVP_aes_128_ocb());
```
It appears so:
```
; g diff
diff --git a/crypto/evp/evp_enc.c b/crypto/evp/evp_enc.c
index b804d74914..d7c7befe1b 100644
--- a/crypto/evp/evp_enc.c
+++ b/crypto/evp/evp_enc.c
@@ -1520,7 +1520,7 @@ void EVP_CIPHER_free(EVP_CIPHER *cipher)
 {
     int i;
 
-    if (cipher == NULL)
+    if (cipher == NULL || cipher->prov == NULL)
         return;
 
     CRYPTO_DOWN_REF(&cipher->refcnt, &i, cipher->lock);
; cat a.c
#include <openssl/evp.h>
int main(void) {
    EVP_CIPHER_free((EVP_CIPHER *)EVP_aes_128_ocb());
    return 0;
}
; gcc -Iinclude a.c libcrypto.a -ldl -lpthread
; ./a.out                                      
;

**Comment by levitte** [comment]
Please do go ahead and try.  I did yesterday, and it appears that I'm not transmitting the issues I encountered well enough.
I suggest enabling asan, that's quite enlightening.

**Comment by richsalz** [comment]
I guess I am not asking the question clearly enough.  Sorry, I don't know how to do it any better.  I understand the "catch-22" is for `EVP_MD_free` because of the multiple ways of creating one.  Does it also hold for `EVP_CIPHER_free` ?

**Comment by richsalz** [comment]
I did this:
```
; CC='ccache clang' ./config enable-asan
; make clean
; make -sj10
; clang  -Iinclude -pthread -m64 -fsanitize=address -fno-omit-frame-pointer -g -Wa,--noexecstack -Qunused-arguments -Wall -c a.c
; clang -pthread -m64 -fsanitize=address -fno-omit-frame-pointer -g -Wa,--noexecstack -Qunused-arguments -Wall a.o libcrypto.a
;  ./a.out
=8166==AddressSanitizer: libc interceptors initialized
|| `[0x10007fff8000, 0x7fffffffffff]` || HighMem    ||
|| `[0x02008fff7000, 0x10007fff7fff]` || HighShadow ||
|| `[0x00008fff7000, 0x02008fff6fff]` || ShadowGap  ||
|| `[0x00007fff8000, 0x00008fff6fff]` || LowShadow  ||
|| `[0x000000000000, 0x00007fff7fff]` || LowMem     ||
MemToShadow(shadow): 0x00008fff7000 0x000091ff6dff 0x004091ff6e00 0x02008fff6fff
redzone=16
max_redzone=2048
quarantine_size_mb=256M
thread_local_quarantine_size_kb=1024K
malloc_context_size=30
SHADOW_SCALE: 3
SHADOW_GRANULARITY: 8
SHADOW_OFFSET: 0x7fff8000
==8166==Installed the sigaction for signal 11
==8166==Installed the sigaction for signal 7
==8166==Installed the sigaction for signal 8
==8166==T0: stack [0x7ffc5559d000,0x7ffc55d9d000) size 0x800000; local=0x7ffc55d9b228
==8166==AddressSanitizer Init done
==8167==Processing thread 8166.
==8167==Stack at 0x7ffc5559d000-0x7ffc55d9d000 (SP = 0x7ffc55d9af78).
==8167==TLS at 0x7f77ac03e000-0x7f77ac03f080.
;
```

That looks to me like it is okay for *ciphers*.

**Comment by richsalz** [comment]
(Also note that I made a change to `EVP_CIPHER_free`, shown in the diff.)

**Comment by levitte** [comment]
> Does it also hold for `EVP_CIPHER_free` ?

Yes.

The problem I found was actually with `EVP_MD_CTX_md()`, which may or may not return a provider backed EVP_MD, but that one should *not* be freed, but if you still do, there's no way for `EVP_MD_free()` to tell, so you end up with a double free.
Now, there also exists a `EVP_CIPHER_CTX_cipher()`, so the same problem applies there.

**Comment by t8m** [comment]
Yes, but I repeat that the md returned from EVP_MD_CTX_md() (and cipher from EVP_CIPHER_CTX_cipher()) should never be handled in the same way as those that you get from algorithm lookup (or fetch). It's practically the same as with other attributes obtained via the get0 functions. So I do not see the existence of those two functions to be a blocker for allowing EVP_MD_free() to be called on the md obtained from the algorithm lookup.

**Comment by levitte** [comment]
> Yes, but I repeat that the md returned from EVP_MD_CTX_md() (and cipher from EVP_CIPHER_CTX_cipher()) should never be handled in the same way as those that you get from algorithm lookup (or fetch). It's practically the same as with other attributes obtained via the get0 functions. So I do not see the existence of those two functions to be a blocker for allowing EVP_MD_free() to be called on the md obtained from the algorithm lookup.

So you still need *two* variables, one extra to hold what was returned from `EVP_MD_CTX_md()`, and some mechanism to know that this one should be used for what follows, but that one must still never be freed by the application.  May I suggest that with that idea, you study this bit of code and try to think of an elegant way to code it so the pointer from the `EVP_MD_CTX_md()` doesn't get freed?  I wasn't able to...

https://github.com/openssl/openssl/blob/68883d9db86534176d744c7691ac7565f5def884/apps/dgst.c#L403-L420

**Comment by t8m** [comment]
Well you'd still have to use another fetched_md variable to be freed here. There is no way around that.

**Comment by levitte** [comment]
> Well you'd still have to use another fetched_md variable to be freed here. There is no way around that.

Thank you.

The other point that I would like to make is that *a public library should **not** encourage users to try to free const things*!
That should be a core tenet in API design, and what's being proposed in #14193 goes against that.

**Comment by t8m** [comment]
Yes, that I completely agree with. We would have to change EVP_getdigestby... to return non-const EVP_MD. Which would be weird.
Anyway I do not want to fight for this change as I do not see the current API contract as too big issue. If new applications do not care about engines support, they are not affected by this complexity at all. And applications that want to deal with both engines and providers can handle that.

**Comment by richsalz** [comment]
I believe this was fixed in #14135 and should be closed.

**Comment by t8m** [comment]
Yes, fetch is now used in both opt_cipher and opt_md.
